03-rnaseq/pathway-analysis_rnaseq_02_gsea.Rmd

```diff
@@ -0,0 +1,556 @@
+
+```{r}
+# First let's create a mapped data frame we can join to the differential expression stats
+dge_mapped_df <- data.frame(
+  "gene_symbol" = mapIds(
+    org.Hs.eg.db, # Replace with annotation package for the organism relevant to your data
+    keys = dge_df$Gene,
+    column = "SYMBOL", # Replace with the type of gene identifiers you would like to map to
+    keytype = "ENSEMBL", # Replace with the type of gene identifiers in your data
+    multiVals = "first" # This will keep only the first mapped value for each Ensembl ID
+  )
+) %>%
+  # If an Ensembl gene identifier doesn't map to a gene symbol, drop that
+  # from the data frame
+  dplyr::filter(!is.na(gene_symbol)) %>%
+  # Make an `Ensembl` column to store the rownames
+  tibble::rownames_to_column("Ensembl") %>%
+  # Now let's join the rest of the expression data
+  dplyr::inner_join(dge_df, by = c("Ensembl" = "Gene"))
+```
+
+This `1:many mapping between keys and columns` message means that some Ensembl gene identifiers map to multiple gene symbols.
+In this case, it's also possible that a gene symbol will map to multiple Ensembl IDs.
+For the purpose of performing GSEA later in this notebook, we keep only the first mapped IDs.
+For more about how to explore this, take a look at our [microarray gene ID conversion example](https://alexslemonade.github.io/refinebio-examples/02-microarray/gene-id-annotation_microarray_01_ensembl.html).
+
+Let's see a preview of `dge_mapped_df`.
+
+```{r}
+head(dge_mapped_df)
+```
+
+We will want to keep in mind that GSEA requires that genes are ranked from most highly positive to most highly negative and weighted according to their gene-level statistic, which is essential to the calculation of the enrichment score (ES), a pathway-level statistic.
+GSEA also requires unique gene identifiers to produce the most accurate results.
+This means if any duplicate gene identifiers are found in our dataset, we will want to retain the instance with the higher absolute value as this will be the instance most likely to be ranked most highly negative or most highly positive.
+Otherwise the enrichment score results may be skewed and the `GSEA()` function will throw a warning.
+
+Let's check to see if we have any gene symbols that mapped to multiple Ensembl IDs.
+
+```{r}
+any(duplicated(dge_mapped_df$gene_symbol))
+```
+
+Looks like we do have duplicated gene symbols.
+Let's find out which gene symbols have been duplicated.
+
+```{r}
+dup_gene_symbols <- dge_mapped_df %>%
+  dplyr::filter(duplicated(gene_symbol)) %>%
+  dplyr::pull(gene_symbol)
+
+dup_gene_symbols
+```
+
+Now let's take a look at the rows associated with the duplicated gene symbols.
+
+```{r}
+dge_mapped_df %>%
+  dplyr::filter(gene_symbol %in% dup_gene_symbols) %>%
+  dplyr::arrange(gene_symbol)
+```
+**IDENTICAL TO MICROARRAY EXAMPLE** (except using absolute log2 fold change value instead of t-statistic value)
```

Comment:
Fold changes could sometimes be deceiving in some instances; you could have one big sample that drives things for some genes. 

What about using `-log10(padj)` instead?  Or `lfcSE` would be better than plain log2 fold change too.

Reply:
I tested whether or not using `lfcSE` rather than `log2FoldChange` would make a difference and there were indeed differences where the data frame filtered to handle duplicate gene identifiers is not identical, and the most positive and most negative pathways according to NES are also not identical.

That said, I visited the `DESeq2` docs and found that the `lfcShrink()` function (that we implemented the RNA-seq differential expression example) is expected to handle instances where one big sample may drive things by _shrinking_ the fold changes to account for genes with very low counts and highly variable counts. So this should not be an issue here.

Seeing as it appears that using `log2FoldChange` is more standard in exploring `DESeq2` results downstream (at least visually as seen in [this RNA-seq workflow vignette](http://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html) and the [`DESeq2` vignette](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) and other smaller examples I found), I am inclined to stick with using `log2FoldChange` here.

I will also note, that when using `lfcSE`, we get the following warning message from the `GSEA()` function:
```
Warning in preparePathwaysAndStats(pathways, stats, minSize, maxSize,
 gseaParam, : All values in the stats vector are greater than zero and scoreType
 is "std", maybe you should switch to scoreType = "pos".
```

which of course can be handled by adding the suggested argument `scoreType = "pos"` but I am not sure is ideal.

For your reference @cansavvy, (and visualization of the varying plots), here is the [rendered html using `lfcSE`](https://cbethell.github.io/html-output/pathway-analysis_rnaseq_02_gsea_lfcse.html) and here is the [rendered html using `log2FoldChange`](https://cbethell.github.io/html-output/pathway-analysis_rnaseq_02_gsea_log2fc.html).

Reply:
>  lfcShrink() function (that we implemented the RNA-seq differential expression example) is expected to handle instances where one big sample may drive things by shrinking the fold changes to account for genes with very low counts and highly variable counts.

Good to know about lfcShrink()! This is good to know!